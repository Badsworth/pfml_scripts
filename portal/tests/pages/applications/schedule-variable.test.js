import {
  DayOfWeek,
  WorkPattern,
  WorkPatternDay,
  WorkPatternType,
} from "../../../src/models/BenefitsApplication";
import {
  MockClaimBuilder,
  renderWithAppLogic,
  simulateEvents,
} from "../../test-utils";
import ScheduleVariable from "../../../src/pages/applications/schedule-variable";

jest.mock("../../../src/hooks/useAppLogic");

const defaultClaim = new MockClaimBuilder()
  .continuous()
  .workPattern({
    work_pattern_days: [],
    work_pattern_type: WorkPatternType.variable,
  })
  .create();

const setup = (claimAttrs = defaultClaim) => {
  const { appLogic, claim, wrapper } = renderWithAppLogic(ScheduleVariable, {
    claimAttrs,
  });

  const { changeField, submitForm } = simulateEvents(wrapper);

  return {
    appLogic,
    changeField,
    claim,
    submitForm,
    wrapper,
  };
};

describe("ScheduleVariable", () => {
  it("renders the form", () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("Trans").dive()).toMatchSnapshot();
  });

  it("submits hours_worked_per_week and 7 day work pattern when entering hours for the first time", async () => {
    const { appLogic, claim, changeField, submitForm } = setup();

    changeField("work_pattern.work_pattern_days[0].minutes", 60 * 7); // 1 hour each day
    await submitForm();

    expect(appLogic.claims.update).toHaveBeenCalledWith(claim.application_id, {
      hours_worked_per_week: 7,
      work_pattern: {
        work_pattern_days: Object.values(DayOfWeek).map(
          (day_of_week) => new WorkPatternDay({ day_of_week, minutes: 60 })
        ),
      },
    });
  });

  it("submits updated data when user changes their answer", async () => {
    const initialWorkPattern = WorkPattern.createWithWeek(60 * 7); // 1 hour each day
    const { appLogic, claim, changeField, submitForm } = setup(
      new MockClaimBuilder()
        .continuous()
        .workPattern({
          work_pattern_days: initialWorkPattern.work_pattern_days,
          work_pattern_type: WorkPatternType.variable,
        })
        .create()
    );

    changeField("work_pattern.work_pattern_days[0].minutes", 2 * 60 * 7); // 2 hour each day
    await submitForm();

    expect(appLogic.claims.update).toHaveBeenCalledWith(claim.application_id, {
      hours_worked_per_week: 14,
      work_pattern: {
        work_pattern_days: Object.values(DayOfWeek).map(
          (day_of_week) => new WorkPatternDay({ day_of_week, minutes: 120 })
        ),
      },
    });
  });

  it("submits data when user doesn't change their answers", async () => {
    const initialWorkPattern = WorkPattern.createWithWeek(60 * 7); // 1 hour each day
    const { appLogic, claim, submitForm } = setup(
      new MockClaimBuilder()
        .continuous()
        .workPattern({
          work_pattern_days: initialWorkPattern.work_pattern_days,
          work_pattern_type: WorkPatternType.variable,
        })
        .create()
    );

    await submitForm();

    expect(appLogic.claims.update).toHaveBeenCalledWith(claim.application_id, {
      hours_worked_per_week: 7,
      work_pattern: {
        work_pattern_days: Object.values(DayOfWeek).map(
          (day_of_week) => new WorkPatternDay({ day_of_week, minutes: 60 })
        ),
      },
    });
  });

  it("creates a blank work pattern when user doesn't enter a time amount", async () => {
    const { appLogic, claim, submitForm } = setup();

    await submitForm();

    expect(appLogic.claims.update).toHaveBeenCalledWith(claim.application_id, {
      hours_worked_per_week: null,
      work_pattern: {
        work_pattern_days: [],
      },
    });
  });
});
