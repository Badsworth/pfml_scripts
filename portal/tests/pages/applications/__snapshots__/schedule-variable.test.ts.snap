// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScheduleVariable renders the form 1`] = `
<div>
  <form
    class="usa-form"
    method="post"
  >
    <h1
      class="js-title margin-top-0 margin-bottom-2 font-heading-sm line-height-sans-3"
      tabindex="-1"
    >
      Employment information
    </h1>
    <h2
      class="font-heading-lg text-bold"
    >
      How many hours do you work on average each week?
    </h2>
    <p
      class="usa-intro"
    >
      Weâ€™ll confirm your hours with your employer after you submit your application. Learn more about 
      <a
        href="https://www.mass.gov/pfml-calculate-hours"
        rel="noopener"
        target="_blank"
      >
        how we use this number
      </a>
       and how to figure out your average.
    </p>
    <fieldset
      class="usa-fieldset margin-top-3 usa-form-group"
    >
      <legend
        class="usa-label text-bold usa-legend font-heading-xs measure-5"
      >
        Average weekly hours
      </legend>
      <div
        class="usa-form-group display-inline-block margin-right-3 margin-top-105"
      >
        <label
          class="usa-label text-normal font-heading-xs measure-5"
          for="InputText1"
          id="InputText1_label"
        >
          Hours
        </label>
        <input
          aria-labelledby="InputText1_label InputText1_hint InputText1_error"
          class="usa-input usa-input--small"
          data-value-type="integer"
          id="InputText1"
          inputmode="numeric"
          name="hours"
          type="text"
          value=""
        />
      </div>
      <div
        class="usa-form-group display-inline-block margin-top-105"
      >
        <label
          class="usa-label text-normal font-heading-xs measure-5"
          for="Dropdown2"
          id="Dropdown2_label"
        >
          Minutes
        </label>
        <div
          class=""
          data-default-value=""
        >
          <select
            aria-labelledby="Dropdown2_label Dropdown2_hint Dropdown2_error"
            class="usa-select maxw-mobile-lg width-15"
            id="Dropdown2"
            name="minutes"
          >
            <option
              value=""
            />
            <option
              value="0"
            >
              0
            </option>
            <option
              value="15"
            >
              15
            </option>
            <option
              value="30"
            >
              30
            </option>
            <option
              value="45"
            >
              45
            </option>
          </select>
        </div>
      </div>
    </fieldset>
    <button
      class="usa-button position-relative margin-top-4"
      type="submit"
    >
      Save and continue
    </button>
  </form>
</div>
`;
